Refactor cabal-install solver config log output

erikd opened the pull request:
Includes:

* Apply some of @grayjay and @mpickering comments
* Fix #4251

This is the PR #9541 rebased and fixed to build.

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [ ] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [ ] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

ulysses4ever commented:
Any chance you could add examples of what the new output looks like? Say, in the PR description.

grayjay commented:
I haven't had a chance to look at the code yet, but, if I remember correctly, #9541 was a followup to #9159. Would it make sense to first update and merge #9159?

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,151 @@ import qualified Data.Set as S
+      where
+        f versionString m = let (pkg, ver) = splitOnLastHyphen versionString
+                            in M.insertWith (++) pkg [ver] m
+        -- FIXME: This is not a very robust way to split the package name and version.
+        -- I should rather retrieve the package name and version from the QPN ...
```
This is going to break on boot package names like `base-4.14.0.0/installed-4.14.0.0`.

erikd replied:
Yep, need to fix that.

erikd replied:
@Bodigrim Can you give an example of it breaking on boot package names? The `cabal-install:unit-tests` tests have instances of names like that and the tests pass fine.

mpickering replied:
Please implement the rendered using structured types. Taking shortcuts is how bugs manifest.

erikd replied:
@mpickering Which types should I be using? I do not know this codebase well.

erikd replied:
Thanks @mpickering ! Changes made.

erikd commented:
> I haven't had a chance to look at the code yet, but, if I remember correctly, #9541 was a followup to #9159. Would it make sense to first update and merge #9159?

According to [this comment](https://github.com/haskell/cabal/pull/9159#issuecomment-1862621844) this seems like a precursor to #9159 .

grayjay commented:
> According to [this comment](https://github.com/haskell/cabal/pull/9159#issuecomment-1862621844) this seems like a precursor to #9159 .

The original change in #9159 was split into a refactoring change and a fix for #4251. Now the refactoring change is in #9159, and the fix for #4251 is in #9541. #9541 contains #9159, because the fix depends on the refactoring.

#9560 has also been merged since #9541 was written and helps address #4251. Do you know how this fix compares now?

erikd commented:
Current version of this PR aims to minimize the differences in the `cabal-install:unit-test` output.

Still need to provide a information about how this version improves the solver output compared to the current output.

erikd commented:
The changes between the output on `master` and the output in this PR is mostly incredibly minor (as shown by the tiny patch to the tests).

This is one of the few differences:
`master`
```
      minimize conflict set with --minimize-conflict-set:                                                                              FAIL
        tests/UnitTests/Distribution/Solver/Modular/DSL/TestCaseUtils.hs:274:
        Unexpected error:
        Could not resolve dependencies:
        [__0] tryingA: A-3.0.0 (user goal)
        [__1] next goal: D (dependency of A)
        [__1] rejectingC: D-1.0.0 (conflict: A => D==2.0.0)
        [__1] fail (backjumping, conflict set: A, D)
        After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: A (5), D (4)
        Use -p '/minimize conflict set with --minimize-conflict-set/' to rerun this test only.
      show original conflict set with --no-minimize-conflict-set:                                                                      FAIL
        tests/UnitTests/Distribution/Solver/Modular/DSL/TestCaseUtils.hs:274:
        Unexpected error:
        Could not resolve dependencies:
        [__0] tryingA: A-3.0.0 (user goal)
        [__1] next goal: B (dependency of A)
        [__1] rejectingC: B-1.0.0 (conflict: A => B==2.0.0)
        [__1] fail (backjumping, conflict set: A, B)
        After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: A (7), B (2), C (2), D (2)
        Try running with --minimize-conflict-set to improve the error message.
        Use -p '/show original conflict set with --no-minimize-conflict-set/' to rerun this test only.
      show full installed package version (issue #5892):                                                                               FAIL
        tests/UnitTests/Distribution/Solver/Modular/DSL/TestCaseUtils.hs:274:
        Unexpected error:
        Could not resolve dependencies:
        [__0] tryingA: my-package-1.0.0 (user goal)
        [__1] next goal: other-package (dependency of my-package)
        [__1] rejectingC: other-package-2.0.0/installed-2.0.0 (conflict: my-package => other-package==3.0.0)
        [__1] fail (backjumping, conflict set: my-package, other-package)
        After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: my-package, other-package
        Use -p '/show full installed package version (issue #5892)/' to rerun this test only.
```
New:
```
      minimize conflict set with --minimize-conflict-set:                                                                              FAIL
        tests/UnitTests/Distribution/Solver/Modular/DSL/TestCaseUtils.hs:268:
        Unexpected solver log:
        targets: A
        constraints: 
          any.base installed (non-reinstallable package)
          any.ghc-bignum installed (non-reinstallable package)
          any.ghc-internal installed (non-reinstallable package)
          any.ghc-prim installed (non-reinstallable package)
          any.ghc installed (non-reinstallable package)
          any.integer-gmp installed (non-reinstallable package)
          any.integer-simple installed (non-reinstallable package)
          any.template-haskell installed (non-reinstallable package)
        preferences: 
        strategy: PreferLatestForSelected
        reorder goals: False
        count conflicts: True
        fine grained conflicts: True
        minimize conflict set: True
        independent goals: False
        avoid reinstalls: False
        shadow packages: False
        strong flags: False
        allow boot library installs: False
        only constrained packages: OnlyConstrainedNone
        max backjumps: infinite
        [__0] tryingE: A-3.0.0 (user goal)
        [__1] next goal: B (dependency of A)
        [__1] rejecting: B-1.0.0 (conflict: A => B==2.0.0)
        [__1] failH (backjumping, conflict set: A, B)
        [__0] tryingE: A-2.0.0
        [__1] tryingE: B-1.0.0 (dependency of A)
        [__2] next goal: C (dependency of A)
        [__2] rejecting: C-1.0.0 (conflict: A => C==2.0.0)
        [__2] failH (backjumping, conflict set: A, C)
        [__0] tryingE: A-1.0.0
        [__1] tryingE: B-1.0.0 (dependency of A)
        [__2] tryingE: C-1.0.0 (dependency of A)
        [__3] next goal: D (dependency of A)
        [__3] rejecting: D-1.0.0 (conflict: A => D==2.0.0)
        [__3] failH (backjumping, conflict set: A, D)
        [__0] failH (backjumping, conflict set: A, B, C, D)
        Found no solution after exhaustively searching the dependency tree. Rerunning the dependency solver to minimize the conflict set ({A, B, C, D}).
        Trying to remove variable "A" from the conflict set.
        [__0] tryingE: A-3.0.0 (user goal)
        [__1] next goal: B (dependency of A)
        [__1] rejecting: B-1.0.0 (conflict: A => B==2.0.0)
        [__1] failH (backjumping, conflict set: A, B)
        [__0] tryingE: A-2.0.0
        [__1] tryingE: B-1.0.0 (dependency of A)
        [__2] next goal: C (dependency of A)
        [__2] rejecting: C-1.0.0 (conflict: A => C==2.0.0)
        [__2] failH (backjumping, conflict set: A, C)
        [__0] tryingE: A-1.0.0
        [__1] tryingE: B-1.0.0 (dependency of A)
        [__2] tryingE: C-1.0.0 (dependency of A)
        [__3] next goal: D (dependency of A)
        [__3] rejecting: D-1.0.0 (conflict: A => D==2.0.0)
        [__3] failH (backjumping, conflict set: A, D)
        [__0] failH (backjumping, conflict set: A, B, C, D)
        Failed to remove "A" from the conflict set. Continuing with {A, B, C, D}.
        Trying to remove variable "B" from the conflict set.
        [__0] tryingE: A-3.0.0 (user goal)
        [__1] tryingE: C-1.0.0 (dependency of A)
        [__2] next goal: D (dependency of A)
        [__2] rejecting: D-1.0.0 (conflict: A => D==2.0.0)
        [__2] failH (backjumping, conflict set: A, D)
        [__0] skippingI: A; 1.0.0, 2.0.0 (has the same characteristics that caused the previous version to fail: depends on 'D' but excludes version 1.0.0)
        [__0] failH (backjumping, conflict set: A, D)
        Successfully removed "B" from the conflict set. Continuing with {A, D}.
        Trying to remove variable "D" from the conflict set.
        [__0] tryingE: A-3.0.0 (user goal)
        [__1] next goal: D (dependency of A)
        [__1] rejecting: D-1.0.0 (conflict: A => D==2.0.0)
        [__1] failH (backjumping, conflict set: A, D)
        [__0] skippingI: A; 1.0.0, 2.0.0 (has the same characteristics that caused the previous version to fail: depends on 'D' but excludes version 1.0.0)
        [__0] failH (backjumping, conflict set: A, D)
        Failed to remove "D" from the conflict set. Continuing with {A, D}.
```
I suppose the main benefit of this PR is that in the file `cabal-install-solver/src/Distribution/Solver/Modular/Message.hs` detection of errors is separated from reporting of errors.

erikd commented:
> I haven't had a chance to look at the code yet, but, if I remember correctly, #9541 was a followup to #9159. Would it make sense to first update and merge #9159?

I have had a look at #9159 (against `master` from 18 months ago) but its rather difficult what really changes between them. I am studying #9159 more closely to figure out what it actually does.

grayjay commented:
> Update: I am not able find any significant differences between the behavior of the code in this PR and in #9541 .

Isn't this PR an updated version of #9541? Do you mean #9560?

erikd commented:
> > Update: I am not able find any significant differences between the behavior of the code in this PR and in #9541 .
> 
> Isn't this PR an updated version of #9541? Do you mean #9560?

Yes, I got confused.  This is an updated version of #9541. Ad for #9560, that has been merged but does not seem to be related this PR.

grayjay commented:
Here is the current state of each of the PRs, as I understand it:

#9159: It contains two commits (b20ea537f7eee5058a50204e2e16856233052613 and f10dbcf7872927622250f25b463bb608882737a9) that refactor the code in preparation for the improvement to error messages in #9541.
 #9541: It contains the two commits from #9159, as well as a third commit (e4775b426c8aac4229509b51f98cc0f024c60478) to improve error messages by removing duplication of package names.
#9560: It removes duplication of package names in error messages, without significant refactoring.

Since #9560 was already merged, it seems like the main difference between this PR and master is the refactoring. Are you interested in merging this just for the refactoring? Are you planning to make additional changes to the solver log that depend on it?

erikd commented:
My hope is to get this refactor merged (after [the fix](https://github.com/haskell/cabal/pull/10854#discussion_r2033116486) suggested by @mpickering ). Then I intend to work on improving error messages as per commit https://github.com/haskell/cabal/commit/e4775b426c8aac4229509b51f98cc0f024c60478 .

grayjay commented:
I meant that this PR already contains the contents of e4775b426c8aac4229509b51f98cc0f024c60478 (removing duplicate package names, similar to #9560), so I was wondering if you were planning to make more changes beyond e4775b426c8aac4229509b51f98cc0f024c60478 that depend on the refactoring.

erikd commented:
@grayjay I had not planned to do any further refactoring after this, but would be happy to be pointed to work that needs to be done.

grayjay commented:
@erikd I feel like I still don't understand the reason for this PR, given that #9560 was already merged. What is your goal in updating #9541?

erikd commented:
> @erikd I feel like I still don't understand the reason for this PR, given that #9560 was already merged. What is your goal in updating #9541?

I think the replacement of a `String` type with `SummarizedMessage` is a significant cosmetic improvement.

If you disagree, please say so an I will close this PR.

grayjay commented:
I think that it's fine to merge just the refactoring, especially just replacing `String` with `SummarizedMessage`. #9159 still has a lot of unaddressed code review comments, though. I could try to copy the relevant ones to this PR if that would help.

erikd commented:
@grayjay Let me look at https://github.com/haskell/cabal/pull/9159 and try to address those comments.

erikd commented:
@grayjay Sorry for the delay.  Here is the updated PR with the change requests make in #9159.

* `LogSucessMsg` -> `LogSuccess`
* `LogFailureMsg` -> `LogFailure`
* All constructors of `Entry` data type ( file `cabal-install-solver/src/Distribution/Solver/Types/Progress.hs`) renamed from `Log*` to `Entry*` .
* `EntryMsg` data type renamed to `EntryAtLevel`.
* Make `SummarizedMsg` type opaque (constructors not exported).
* Undo some formatting changes for imports

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -1,9 +1,14 @@
+    summarizeMessages,
+    renderSummarizedMessage,
+
+    -- Requird by the doctests, but not used elsewhere so export this
```
```suggestion
    -- Required by the doctests, but not used elsewhere so export this
```

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,132 @@ import qualified Data.Set as S
+displayMessage (EntryFailure c fr) = "fail" ++ showFR c fr
+displayMessage (EntrySkipMany qsn b cs) = "skipping: " ++ showOptions qsn b ++ " " ++ showConflicts cs
+-- Instead of displaying `aeson-1.0.2.1, aeson-1.0.2.0, aeson-1.0.1.0, ...`,
+-- the following line aim to display `aeson: 1.0.2.1, 1.0.2.0, 1.0.1.0, ...`.
```
```suggestion
-- the following line aims to display `aeson: 1.0.2.1, 1.0.2.0, 1.0.1.0, ...`.
```

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,132 @@ import qualified Data.Set as S
+displayMessage (EntryTryingS qsn b) = "trying: " ++ showQSNBool qsn b
+displayMessage (EntryUnknownPackage qpn gr) = "unknown package: " ++ showQPN qpn ++ showGR gr
+displayMessage EntrySuccess = "done"
+displayMessage (EntryFailure c fr) = "fail" ++ showFR c fr
```
```suggestion
displayMessage (EntryFailure c fr) = "fail " ++ showFR c fr
```
...unless `showFR` prepends a space itself.

erikd replied:
`showFR` does prepend a space character.

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,132 @@ import qualified Data.Set as S
+showQPNConflictSet qpn popts =
+  case popts of
+    [] -> ""
+    [x] -> showQPNPOpt qpn x
+    (x:xs) -> showQPNPOpt qpn x ++ ", " ++ L.intercalate ", " (map (\(POption i _) -> showI i) xs)
```
```suggestion
    x:xs -> L.intercalate ", " (showQPNPOpt qpn x : map (\(POption i _) -> showI i) xs)
```

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -98,40 +153,31 @@ showMessages = go 0
 showConflicts :: Set CS.Conflict -> String
 showConflicts conflicts =
-    " (has the same characteristics that caused the previous version to fail: "
+    "(has the same characteristics that caused the previous version to fail: "
```
Are we sure that the leading space was redundant? Is it the case that all usages of `showConflicts` supply a space in between themselves?

erikd replied:
All uses of `showConflicts` adds a space before calling `showConflicts`.

Bodigrim commented on cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs:
```diff
@@ -961,7 +961,7 @@ tests =
                     ]
                   rejecting = "rejecting: A-3.0.0/installed-3.0.0"
-                  skipping = "skipping: A; 2.0.0/installed-2.0.0, 1.0.0/installed-1.0.0"
+                  skipping = "skipping: A; 1.0.0/installed-1.0.0, 2.0.0/installed-2.0.0 "
```
What's the source of the trailing space?

erikd replied:
Its un-needed. Removed it

Bodigrim commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,132 @@ import qualified Data.Set as S
+showQPNConflictSet qpn popts =
+  case popts of
+    [] -> ""
+    [x] -> showQPNPOpt qpn x
```
Isn't this case redundant now?

sebright reviewed (commented):
@erikd No problem! Thanks for making the renaming changes. I tried to copy the remaining comments from #9159, though I may have applied some incorrectly since they were hard to view in the Github UI. I also added a few new comments.

erikd commented on cabal-install-solver/src/Distribution/Solver/Types/Progress.hs:
```diff
@@ -47,3 +59,32 @@ instance Applicative (Progress step fail) where
+  | EntryRejectS QSN Bool ConflictSet FailReason
+  | EntrySkipping (Set CS.Conflict)
+  | EntryTryingF QFN Bool
+  | EntryTryingP QPN POption (Maybe (GoalReason QPN))
```
Done!

erikd replied:
Done!

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -1,9 +1,14 @@
+    summarizeMessages,
+    renderSummarizedMessage,
+
+    -- Required by the doctests, but not used elsewhere so export this
+    -- to avoid compiler warnings.
+    showOptions,
```
Removed the export.

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -1,9 +1,14 @@
 module Distribution.Solver.Modular.Message (
     Message(..),
-    showMessages
+    SummarizedMessage(..),
```
Removed export.

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,82 +19,132 @@ import qualified Data.Set as S
+--
+displayMessage (EntryRejectMany qpn is c fr) = "rejecting: " ++ showQPNConflictSet qpn (reverse is) ++ showFR c fr
+
+showQPNConflictSet :: QPN -> [POption] -> String
+showQPNConflictSet qpn popts =
```
Yes, done!

erikd replied:
```diff
diff --git a/cabal-install-solver/src/Distribution/Solver/Modular/Message.hs b/cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
index 5421787180..de1911289f 100644
--- a/cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
+++ b/cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
@@ -95,13 +95,7 @@ displayMessage (EntrySkipMany qsn b cs) = "skipping: " ++ showOptions qsn b ++ "
 -- Instead of displaying `aeson-1.0.2.1, aeson-1.0.2.0, aeson-1.0.1.0, ...`,
 -- the following line aims to display `aeson: 1.0.2.1, 1.0.2.0, 1.0.1.0, ...`.
 --
-displayMessage (EntryRejectMany qpn is c fr) = "rejecting: " ++ showQPNConflictSet qpn is ++ showFR c fr
-
-showQPNConflictSet :: QPN -> [POption] -> String
-showQPNConflictSet qpn popts =
-  case popts of
-    [] -> ""
-    x:xs -> L.intercalate ", " (showOption qpn x : map (\(POption i _) -> showI i) xs)
+displayMessage (EntryRejectMany qpn is c fr) = "rejecting: " ++ showOptions qpn is ++ showFR c fr
 
 -- | Transforms the structured message type to actual messages (SummarizedMessage s).
 --
```

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -270,6 +293,12 @@ showOptions q xs = showQPN q ++ "; " ++ (L.intercalate ", "
   | x@(POption i linkedTo) <- xs
   ])
 
+showQPNPOpt :: QPN -> POption -> String
+showQPNPOpt qpn@(Q _pp pn) (POption i linkedTo) =
```
Yes, done!

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular.hs:
```diff
@@ -258,7 +278,7 @@ tryToMinimizeConflictSet runSolver sc cs cm =
         continueWith ("Trying to remove variable " ++ varStr ++ " from the "
                       ++ "conflict set.") $
-        retry (runSolver sc') $ \case
+        retry (retryMap renderSummarizedMessage $ runSolver sc') $ \case
```
Removing the `retryMap renderSummarizedMessage` part results in:
```
src/Distribution/Solver/Modular.hs:281:16: error: [GHC-83865]
    • Couldn't match type ‘SummarizedMessage’ with ‘[Char]’
      Expected: RetryLog String SolverFailure a
        Actual: RetryLog SummarizedMessage SolverFailure a
```
This needs further investigation.

erikd replied:
The code is so convoluted. There are numerous places where it converts back and forth between `String` and `SummarizedMessage` :facepalm:

grayjay replied:
I didn't have a chance to look into this very thoroughly, but I think that the issue could be that #9159 hadn't converted all of the code to use `SummarizedMessage` yet. For example, `tryToRemoveOneVar` and this lambda still use strings. I think that the solver should use `SummarizedMessage` everywhere, except for one call to `renderSummarizedMessage` at the end.

erikd replied:
FInally managed to do it. `retryMap` is gone forever!

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular.hs:
```diff
@@ -297,6 +317,9 @@ tryToMinimizeConflictSet runSolver sc cs cm =
         ExhaustiveSearch cs' cm' -> f cs' cm'
         BackjumpLimitReached     -> fromProgress (Fail BackjumpLimitReached)
 
+    retryMap :: (t -> step) -> RetryLog t fail done -> RetryLog step fail done
+    retryMap f l = fromProgress $ (\p -> foldProgress (\x xs -> Step (f x) xs) Fail Done p) $ toProgress l
```
Had a couple of runs at this and was not able to write `retryMap` without using `toProgress` and `fromProgess`.

I am not saying its not possible, just that I am not smart enough to do it. It might even be easy, but I do not see it.

grayjay commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,124 +14,166 @@ import qualified Data.Set as S
+displayMessage (EntryRejectS qsn b c fr) = "rejecting: " ++ showQSNBool qsn b ++ showFR c fr
+displayMessage (EntrySkipping cs) = "skipping: " ++ showConflicts cs
+displayMessage (EntryTryingF qfn b) = "trying: " ++ showQFNBool qfn b
+displayMessage (EntryTryingP qpn i) = "trying: " ++ showOptions qpn [i]
+displayMessage (EntryTryingNewP qpn i gr) = "trying: " ++ showOptions qpn [i] ++ showGR gr
```
I think that both of these lines should use `showOption` instead of `showOptions`, since they both show single versions.

erikd replied:
Done!

erikd replied:
Fixed.

ulysses4ever commented:
Try rebasing on top of current master (if not already).

sebright reviewed (commented):
I responded to some of your latest comments and changes, but I haven't done another full review yet.

erikd commented:
> Try rebasing on top of current master (if not already).

It says:
```
Current branch erikd/cosmetic-changes-2 is up to date.
```

ulysses4ever commented:
Let me try to restart the failing jobs then. It's the same test in all configuration, so this test may be up for the flaky marker if it fails like that ...

erikd commented on cabal-install-solver/src/Distribution/Solver/Types/Progress.hs:
```diff
@@ -47,3 +59,32 @@ instance Applicative (Progress step fail) where
+  | Success
+  | Failure ConflictSet FailReason
+
+data Entry
```
Done.

erikd replied:
Undid `Message` to `SolverState` rename. Back to being named `Message` and in the file `../Solver/Modular/Message.hs`.

erikd commented on cabal-install-solver/src/Distribution/Solver/Types/Progress.hs:
```diff
@@ -47,3 +59,32 @@ instance Applicative (Progress step fail) where
+  | Next (Goal QPN)
+  | Skip (Set CS.Conflict)
+  | Success
+  | Failure ConflictSet FailReason
```
Move back.

ulysses4ever commented:
I'll try to patch up the failing test (it's the same one again) today.

ulysses4ever commented:
I restarted the bootstrap job, but it failed again in the same way: with `http.client.IncompleteRead: IncompleteRead(3307 bytes read)`. https://github.com/haskell/cabal/actions/runs/15456775662/job/43546021620?pr=10854 So, network problems in GitHub runners strike again. I feel like GH is forcing us to make the testsuite more hermetic, which is a great thing!

ulysses4ever reviewed (dismissed):
This PR needs a changelog entry.

Also, the PR description is missing a self-contained description of the change — please add one. I'd expect an example of solver output after this change.

erikd commented:
> I'd expect an example of solver output after this change.

Example solver output is in: https://github.com/haskell/cabal/pull/10854#issuecomment-2771197952

Solver output has not changed since that.

grayjay commented:
I'm confused now. I had thought that we agreed that this PR was just a refactoring change to add an intermediate log message type (`SummarizedMessage`) to the processing of the solver log. I thought that all of the behavior changes from #9541 were already merged in #9560 and that any remaining changes to strings in the solver unit tests were just bugs that needed to be worked out.

Was the change in https://github.com/haskell/cabal/pull/10854#issuecomment-2771197952 automatically caught by tests, or did you need to manually compare the solver output?

geekosaur commented:
> the new output is longer than the old one. It's the opposite of what most people I talk to want

I think longer is fine as long as it's _more comprehensible_. That's the real problem most people have with the current solver output, it's a lot of gobbledygook that requires someone who has some idea of what's going on inside the solver to interpret.

erikd commented:
> Was the change in [#10854 (comment)](https://github.com/haskell/cabal/pull/10854#issuecomment-2771197952) automatically caught by tests, or did you need to manually compare the solver output?

I fiddle with one of the tests (do not even remember which one) to force a failure. The solver output was obviously different.

grayjay commented:
> I fiddled with one of the tests (do not even remember which one) to force a failure. The solver output was obviously different.

It would help to add a unit test that can detect that change, especially if it's unrelated to the changes caught by the unit tests that had changes to their expected strings.

I didn't look at the specific unit test in that comment, but the output from the first one ("master") looks like the summarized log, which is shown at normal verbosity, and the second one ("new") looks like the part of the log that is only shown at higher verbosity (-v3). Did you change the verbosity when running the tests, or is it possible that this PR changes which lines of the solver log are shown at different verbosities?

erikd commented on cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs:
I have managed to fix all the differences except the first one. That is buried somewhere deep in the changes.

erikd commented:
Had to rebase this as chasing down changes was becoming too difficult.

In the validate test there is now a SINGLE difference:
```diff
diff --git a/cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs b/cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs
index a1f5eed3c6..0136e7cacf 100644
--- a/cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs
+++ b/cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs
@@ -208,7 +208,7 @@ tests =
               solverSuccess [("base", 1), ("ghc-prim", 1), ("integer-gmp", 1), ("integer-simple", 1)]
       , runTest $
           mkTest dbNonupgrade "Refuse to install newer ghc requested by another library" ["A"] $
-            solverFailure (isInfixOf "rejecting: ghc-2.0.0 (constraint from non-reinstallable package requires installed instance)")
+            solverFailure (isInfixOf "rejecting: ghc-1.0.0/installed-1 (conflict: A => ghc==2.0.0)")
       ]
   , testGroup
       "reject-unconstrained"
```
that I have been able to track down.

I was not able to find the difference that I found before. I tried subtle modifications to numerous of the validate tests and was comparing the difference between `cabal-3.14` and the cabal build from my changes. I was not even able to find the change I note above.

erikd commented:
:sigh: The changes require to remove most of the changes to the validate test expected results causes the doctests to fail.

grayjay commented:
Which doctests failed? Do you mean the ones in `Message.hs`? Those are just testing an internal function, so I think that it would be better to update the doctests than change the solver's output.

grayjay commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -14,26 +14,51 @@ import qualified Data.Set as S
+
+import Text.PrettyPrint ( nest, render )
 
+-- A data type to hold state information for the modular solver.
 data Message =
```
Similarly to my previous comment, I think that it is more accurate to describe this type as a log message than solver state.

grayjay commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -98,40 +159,32 @@ showMessages = go 0
         goPReject l qpn (i : is) c fr ms
     goPReject l qpn is c fr ms =
-        (atLevel l $ blurbOptions Rejecting qpn (reverse is) ++ showFR c fr)
-        (go l ms)
+        Step (SummarizedMsg $ AtLevel l $ (EntryRejectMany qpn is c fr)) (go l ms)
```
I think that the unit test failures could have been caused by the removal of the call to `reverse` here and in `goPSkip`.

erikd replied:
Reinstating these `reverse` opertations as well as replacing `showQPNConflictSet` with `showOptions` seems to have fixed this. I now have zero changes with respect to `master` in the file cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs

sebright reviewed (commented):
I did another full review. I have some followup comments, and I saw that there are two unresolved comments above (documentation for new types, and removal of the `EntrySkipping` constructor).

erikd commented on cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL/TestCaseUtils.hs:
```diff
@@ -265,7 +265,7 @@ runTest SolverTest{..} = askOption $ \(OptionShowSolverLog showSolverLog) ->
         printMsg msg = when showSolverLog $ putStrLn msg
         msgs = foldProgress (:) (const []) (const []) progress
-    assertBool ("Unexpected solver log:\n" ++ unlines msgs) $
+    assertBool ("Unexpected error:\n" ++ unlines msgs) $
```
Reverted.

erikd commented on cabal-install-solver/src/Distribution/Solver/Types/Progress.hs:
Sorry, I have completel lost track of what that is even referring to.

sebright reviewed (commented):
@erikd Thanks for keeping the commits separate for review. I've had a lot of trouble finding the unresolved comments in the Github UI for this PR, too. I looked through the comments again, and I think these are the only remaining ones, including one new one:

- Removing the `EntrySkipping` constructor:

  From #9159: I think it is a little confusing to have a separate ~`Log`~ `Entry` constructor for a case that should never happen (`EntrySkipping`). Can this case just use `ErrorMsg`?

- Documentation for the new types in `SummarizedMessage.hs`:

  From #9159: It would help to have some comments describing how the new types and `Message` relate to each other.

- Reverting the changes to `Progress.hs`

- Reverting the renaming of `createErrorMsg`, `rerunSolverForErrorMsg`, and `finalErrorMsg`

yvan-sraka committed:
Refactor cabal-install solver config log output

The main goal is to add an intermediate log message type to
the processing of the solver log. There are zero known
changes to the cabal solver's output.

Co-Authored-By: Erik de Castro Lopo <erikd@mega-nerd.com>

sebright reviewed (commented):
> > ```
> > * Removing the `EntrySkipping` constructor:
> > ```
> ...
> 
> There are `skipping: ....` messages in `cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs` but as I pointed out above, none of the `EntrySkipping` renderings are in the unit tests.

I realized that this issue is more involved than I initially thought, and there may be other impossible cases in `summarizeMessages`, too, so it doesn't make sense to handle the issue in this PR. I'm fine with leaving the `EntrySkipping` constructor.

> > ```
> > * Documentation for the new types in `SummarizedMessage.hs`:
> > ```
> 
> Have added documentation but as I did not actually write this code I can only guess the original intent. Suggestions for improvement welcome.

I added a few suggestions.

sebright reviewed (approved):
Thanks! Looks good!

erikd commented:
Rebased into a single commit. Checked that the diff between the rebased version of the previous commit was zero length.

grayjay commented:
Great! Could you please remove the part of the commit message about changing the solver's output and fixing #4251, now that it only contains the refactoring?

grayjay commented:
@erikd Thanks! I think this is ready for a merge label.

erikd commented:
Rebased against master.

erikd commented:
The CI failure is nothing to do with this commit.

geekosaur commented:
That's a "shouldn't happen". Hackage has been hiccuping since the [planned maintenance window](http://status.haskell.org/pages/maintenance/537c07b0cf1fad5830000093/68671ada71848544434d2462) ended.
